File: Algoritm by CPP/Algoritm by CPP/Stack/2504.cpp
//
//  main.cpp
//  Algoritm by CPP
//
//  Created by zaehorang on 2024/02/29.
//  https://www.acmicpc.net/problem/2504

#include <bits/stdc++.h>
using namespace std;

stack<char> s;

int main(void){
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    string str;
    cin >> str;
    
    int ans = 0;
    int n = 1;
    
    for(int i = 0; i < str.size(); i++){
        if(str[i] == '('){
            n *= 2;
            s.push(str[i]);
        }
        else if(str[i] == '['){
            n *= 3;
            s.push(str[i]);
        } 
        else if(str[i] == ')'){
            if(s.empty() || s.top() != '('){
                cout << 0;
                return 0;
            }
            if(str[i-1] == '(') ans += n;
            n /= 2;
            s.pop();
        }
        else{ // ']'
            if(s.empty() || s.top() != '['){
                cout << 0;
                return 0;
            }
            if(str[i-1] == '[') ans += n;
            n /= 3;
            s.pop();
        }
    }
    if(s.empty()) cout << ans;
      else cout << 0;
}

File: Algorithm by Swift/Algorithm by Swift/Stack/2504.swift
//
//  main.swift
//  Algorithm by Swift
//
//  Created by zaehorang on 2024/02/29.
//  https://www.acmicpc.net/problem/2504

let str = readLine()!

var s = [Character]()

var ans = 0
var n = 1  // 곱해지는 값

var prev: Character = " "  // 반복문의 직전 문자 저장

for c in str {
    if c == "(" {
        n *= 2
        s.append(c)
        
    } else if c == "[" {
        n *= 3
        s.append(c)
        
    } else if c == ")" {
        
        if s.isEmpty || s.last != "(" {
            s.append(" ")  // 정답 출력 시, 스택의 빈배열 분기처리로 정답을 한 번만 출력하기 위함
            break
            
        } else if prev == "(" {
            ans += n
            n /= 2
            s.removeLast()
            
        } else { // 경우: 닫 -> 닫
            s.removeLast()
            n /= 2
        }
        
    } else {  // c == "]'
        if s.isEmpty || s.last != "[" {
            s.append(" ")
            break
        } else if prev == "[" {
            ans += n
            n /= 3
            s.removeLast()
        } else { // 경우: 닫 -> 닫
            s.removeLast()
            n /= 3
        }
    }
    
    prev = c  // 직전 문자 기록
}

if s.isEmpty {
    print(ans)
} else {
    print(0)
}
